### PR TITLE
Sema: allow field access instructions to be emitted at comptime

### PR DESCRIPTION
**Pull request by mlugg**
Related to 34e4b07. This makes `struct.arr.len` succeed at comptime.

Resolves: #15280

### CODE REVIEW

**Comment by mlugg** [comment]
Going to extend to a few other things (tuples, unions, arrays)

**Comment by jayschwa** [comment]
#14819 changed some miscellaneous length accesses to magic numbers. It might be good to revert those changes if this PR re-enables the behavior.

Example: `addr.un.path.len + 1` was changed to `200`: https://github.com/ziglang/zig/commit/ccf670c2b04ebeb9db43eb9f5c47c6cf03e4b1d0#diff-902e94d5dd402cc20a4baa94f9b1e51b95e09437b51ca51ede4a436723eee9fbL102-R102

**Comment by mlugg** [comment]
> #14819 changed some miscellaneous length accesses to magic numbers. It might be good to revert those changes if this PR re-enables the behavior.

Good shout - done.

@ reviewers: regarding the second commit, it turned out that doing this validation everywhere made a lot of compile errors worse. This makes sense, since it's more useful for e.g. `switch` expressions to be able to add a note telling you that switch prongs have to be comptime-known. I opted to split `block_comptime` into two tags: one which validates its result to be comptime-known, and one which defers that responsibility to the use site. If that isn't satisfactory, I can drop that commit for now, since it's not necessary for the first, it was just something I noticed while working on it.

File: src/Sema.zig [hunk 1]
```diff
@@ -24126,6 +24154,8 @@ fn tupleFieldValByIndex(
     field_index: u32,
     tuple_ty: Type,
 ) CompileError!Air.Inst.Ref {
+    _ = src;
```
Comment (Vexu) [inline src/Sema.zig:24157]: Remove this parameter.

**Review by Vexu — APPROVED** [review]
I don't have a suggestion for a better name but the current one didn't immediately make it clear that the difference comes from explicitly using the comptime syntax.

**Comment by andrewrk** [comment]
I'd like to give this another look - you mind rebasing and conflict fixing, @mlugg?

**Comment by mlugg** [comment]
Sure thing. I may be a bit busy over the next couple of days with work stuff, but I'll put this on my list and try to get it done within the next week or so!

**Comment by andrewrk** [comment]
Closing abandoned PR

**Pull request closed without merging**